### PR TITLE
[JENKINS-26380] Terminate nodes correctly

### PR DESCRIPTION
**Pull request by stephenc**
[https://issues.jenkins-ci.org/browse/JENKINS-26380](JENKINS-26380)

Supersedes #1.

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/durabletask/executors/OnceRetentionStrategy.java [hunk 1]
```diff
@@ -76,15 +82,38 @@ private void done(Executor executor) {
...
-            c.getNode().terminate();
-        } catch (InterruptedException x) {
-            LOGGER.log(Level.WARNING, null, x);
-        } catch (IOException x) {
-            LOGGER.log(Level.WARNING, null, x);
+        synchronized (this) {
+            if (terminating) {
+                return;
+            }
+            terminating = true;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/durabletask/executors/OnceRetentionStrategy.java]: Seems like an `AtomicBoolean` would be more idiomatic, but perhaps you need the monitor on `OnceRetentionStrategy.this` anyway to be exclusive with the `threadPoolForRemoting` thread?

**Comment by jglick** [comment]
Do not really follow the implications of the change, but seems OK to merge to me.

**Comment by jglick** [comment]
Is this intended to be a final fix of JENKINS-26380, or do you expect other retention strategies are also in danger from the core locking change?

File: src/main/java/org/jenkinsci/plugins/durabletask/executors/OnceRetentionStrategy.java (hunk 1 above)
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/durabletask/executors/OnceRetentionStrategy.java]: correct we want the lock here to ensure exclusivity of starting the background task

**Comment by stephenc** [comment]
@jglick other retention strategies are already broken if they didn't acquire the lock on Queue before checking idle and then in the background repeat after reacquiring the lock on Queue. The core change makes it easier to do that.

It does not make sense to add the lock to Jenkins.removeNode as there may be valid cases where you want to remove the node and jobs in progress be damned... hence it is up to the caller to ensure they have the correct lock based on whether the node being idle is important to them or not

**Pull request merged**